packages/styled-components/src/utils/flatten.js

```diff
@@ -8,45 +8,47 @@ import isStyledComponent from './isStyledComponent';
 import Keyframes from '../models/Keyframes';
 import hyphenate from './hyphenateStyleName';
 import addUnitIfNeeded from './addUnitIfNeeded';
+import { type Stringifier } from '../types';
 
 /**
  * It's falsish not falsy because 0 is allowed.
  */
 const isFalsish = chunk => chunk === undefined || chunk === null || chunk === false || chunk === '';
 
 export const objToCssArray = (obj: Object, prevKey?: string): Array<string | Function> => {
-  const rules = []
-  const keys = Object.keys(obj)
-
-  keys
-    .forEach(key => {
-      if (!isFalsish(obj[key])) {
-        if (isPlainObject(obj[key])) {
-          rules.push(...objToCssArray(obj[key], key))
-
-          return rules
-        }
-        else if (isFunction(obj[key])) {
-          rules.push(`${hyphenate(key)}:`, obj[key], ';')
-
-          return rules
-        }
-        rules.push(`${hyphenate(key)}: ${addUnitIfNeeded(key, obj[key])};`);
+  const rules = [];
+  const keys = Object.keys(obj);
+
+  keys.forEach(key => {
+    if (!isFalsish(obj[key])) {
```

Comment:
- can we get rid of the nested statements?
- why are we returning `rules`?
- should we just use for-in and `hasOwnProperty` here?

Reply:
This actually made a massive difference in the performance of our object style implementation! Awesome review